### DISCUSSION TITLE
ENCODER sbc ERROR

### DISCUSSION THREAD

**Discussion by aqildad-create** [pschatzmann/arduino-audio-tools#742] [Q&A] [answered]
After some diagnose and trouble shooing i observe below line through exception , 
AudioInfo to(44100, 2, 16);
SBCEncoder encoder;
.
encoder.begin(to);
.
.
.
static uint8_t buffer[512];
encoder.write(buffer, 512); //.>>>>>>>>>>>>>>> THIS LINE GIVE EXCEPTION ERROR 

BELOW IS COMPLETE CODE ,

https://github.com/pschatzmann/arduino-audio-tools/issues/738

**Comment by pschatzmann** [comment, marked as answer]
You should probably only use Encoders directly if you know what you are doing and if you are able to analyse errors in order to figure out what you are doing wrong. My first rough guess is that you just forgot to indicate the output.
So please stick with the examples using the EncodedAudioStream...

If you find any errors, please feel free to correct them and share the correction with a pull request.

**Comment by aqildad-create** [comment]
File EncodedDataFile;
SBCEncoder encoder(EncodedDataFile);

In complete code I write encoded data to file , I think this is encoder out
put .it also through same exemption which image I have uploaded .

Further for this test I follow your other example which is without encoded
stream .. I think this should also work ..

**Reply by pschatzmann** [reply]
What are you talking about ? Look at the class documentation of the SBCEncoder! 
The constructor is asking for SBC parameters!

**Comment by aqildad-create** [comment]
this is the minimal code which reproduce this issue  
 is any parameter still missing ? 


#include "AudioTools.h"
#include "AudioCodecs/CodecSBC.h"
AudioInfo to(44100, 2, 16);
SBCEncoder encoder;
void setup(void) {
    Serial.begin(115200);
    AudioLogger::instance().begin(Serial, AudioLogger::Info);
    encoder.begin(to);
    static uint8_t buffer[512];
    encoder.write(buffer, 512);
}
void loop() {
    
}

![image](https://user-images.githubusercontent.com/65456998/229315353-c836bd40-81cf-4551-a870-c967853ea2da.png)

**Reply by pschatzmann** [reply]
You did not specify the output: Look at the class documentation how to do this!

**Reply by aqildad-create** [reply]
I am extremely apology for annoying question.

I am in  class documentation but i cant figure out  how to specify the out put in above minimal code please .

![image](https://user-images.githubusercontent.com/65456998/229315750-fbfb4c88-fd33-4c79-a485-2e67ffcacd1b.png)

**Reply by pschatzmann** [reply]
Use the setOutputStream() method!

**Comment by aqildad-create** [comment]
ok some thing like this ? and see the out put 75 bytes have data its means 512 bytes of uncompressed data compressed to 75 bytes ? 

#include "AudioTools.h"
#include "AudioCodecs/CodecSBC.h"
AudioInfo to(44100, 2, 16);
SBCEncoder encoder;
RingBufferStream  outdata;
void setup(void) {
    Serial.begin(115200);
    AudioLogger::instance().begin(Serial, AudioLogger::Info);
    encoder.begin(to);
    encoder.setOutputStream(outdata);
    static uint8_t buffer[512];
    encoder.write(buffer, 512);
    outdata.readBytes(buffer, 512);
    for(int aa=0;aa<512;aa++)  // view  compressed data 
    Serial.println(String(aa)+":"+buffer[aa]);
    
    }
void loop() {
    
}

//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
[I] CodecSBC.h : 247 - virtual void audio_tools::SBCEncoder::begin()
0:156
1:185
2:32
3:202
4:0
5:0
6:0
7:0
8:0
9:0
10:0
11:0
12:85
13:85
14:85
15:85
16:85
17:85
18:85
19:85
20:85
21:85
22:85
23:85
24:85
25:85
26:85
27:85
28:85
29:85
30:85
31:85
32:85
33:85
34:85
35:85
36:85
37:85
38:85
39:85
40:85
41:85
42:85
43:85
44:85
45:85
46:85
47:85
48:85
49:85
50:85
51:85
52:85
53:85
54:85
55:85
56:85
57:85
58:85
59:85
60:85
61:85
62:85
63:85
64:85
65:85
66:85
67:85
68:85
69:85
70:85
71:85
72:85
73:85
74:85
75:85
76:0
77:0
78:0
79:0
80:0
81:0

**Reply by pschatzmann** [reply]
It depends on the parameters. You can query the API for the input and output record size
